pubsub/gcppubsub: Allow configuration of batch sizes

davidsbond opened the pull request:
Fixes #3003 

This PR allows users of the `pubsub/gcppubsub` package to configure the receive batch size via a query parameter. 

Signed-off-by: David Bond <davidsbond93@gmail.com>

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,63 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+			}
+
+			o.SubscriptionOptions.AckMaxHandlers = int(maxHandlers)
+		case "ack-min-batch-size":
```
I don't think we need to make the ack batchers configurable. They should be driven by the receive batches.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,63 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+			}
+
+			o.SubscriptionOptions.ReceiveMaxHandlers = int(maxHandlers)
+		case "receive-min-batch-size":
```
I don't think we should make this configurable. What's the use case? The concrete class already starts at `1` and dynamically ramps up, it's not obvious to me how a default other than 1 would behave.

davidsbond replied:
I mostly went for "If I'm configuring one, I can configure them all". But I'll omit them all besides the max batch size

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,63 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+
+	for param, value := range u.Query() {
+		switch param {
+		case "receive-max-handlers":
```
Not sure this needs to be configurable either, is there a use case?

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,63 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+			}
+
+			o.SubscriptionOptions.ReceiveMinBatchSize = int(minBatchSize)
+		case "receive-max-batch-size":
```
Need to ensure that it's not > 1000, since GCP doesn't support anything larger.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,63 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
-		return nil, fmt.Errorf("open subscription %v: invalid query parameter %q", u, param)
+	// Set subscription options to use defaults
+	o.SubscriptionOptions = SubscriptionOptions{
+		ReceiveMaxHandlers:  recvBatcherOpts.MaxHandlers,
```
If we do this, let's rename `recvBatchOpts` -> `defaultRecvBatchOpts` to make clear it's not used explicitly.

In fact, if you do that you fill find that it *is* used explicitly, on line 519; that will need to be updated to use the saved `SubscriptionOptions`.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -392,14 +448,34 @@ type subscription struct {
 // SubscriptionOptions will contain configuration for subscriptions.
-type SubscriptionOptions struct{}
+type SubscriptionOptions struct {
+	ReceiveMaxHandlers  int
```
All of these that remain would need to be documented.

Some of them are not easy to explain because they are used in non-trivial ways by the concrete implementation.

I think we can keep only `MaxBatchSize` (no need for the `Receive` prefix), with the docstring something like:

"MaxBatchSize caps the maximum batch sized used when retrieving messages. It defaults to 1000."

davidsbond commented:
@vangent Thanks for taking the time to look at my PR. I've implemented your feedback and also added some extra `nil` checks after seeing errors in the tests suite.

davidsbond commented:
It also seems like the project has ran out of credits for TravisCI:

https://travis-ci.com/github/google/go-cloud/requests

So the build hasn't run

davidsbond commented:
Did a force push to get the Travis CI to run, seems there's credits again now

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,27 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
-	for param := range u.Query() {
-		return nil, fmt.Errorf("open subscription %v: invalid query parameter %q", u, param)
+	// Set subscription options to use defaults
+	o.SubscriptionOptions = SubscriptionOptions{
```
This doesn't look right, because 
a) it will override any `SubscriptionOptions` that the user actually provided if they are using a custom `Opener`.
b) Down below, the struct is updated based on the URL parameter; since we pass a pointer to the struct to `OpenSubscription`, that essentially updates all of the subscribers to whatever the latest URL parameter was.

I think this should clone the struct and update the `MaxBatchSize` based on the URL parameter. There's no need to worry about the default here since it is handled later.

Something like:

```
opts := o.SubscriberOptions // clone the options
```

and then update `ops.MaxBatchSize` down below as well.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,27 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+				return nil, fmt.Errorf("open subscription %v: invalid query parameter %q: must be between 1 and 1000", u, param)
+			}
+
+			o.SubscriptionOptions.MaxBatchSize = int(maxBatchSize)
```
Is the cast to `int` needed here?

davidsbond replied:
Nope, I'll make the function use `strconv.Atoi` instead as suggested. Not sure why I forgot about that function to begin with.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -210,8 +211,27 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
+
+	for param, value := range u.Query() {
+		switch param {
+		case "max-batch-size":
```
Please update the comments near line 182 to document this URL parameter.

See https://github.com/google/go-cloud/blob/master/blob/gcsblob/gcsblob.go#L204 for an example. Make sure it's clear that it only affects Subscriber.

vangent replied:
Please modify this to `max_recv_batch_size` to match the style used elsewhere in the Go CDK, and to make it clear that it's specific to subscriptions.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -387,19 +407,34 @@ func (*topic) ErrorCode(err error) gcerrors.ErrorCode {
 // SubscriptionOptions will contain configuration for subscriptions.
-type SubscriptionOptions struct{}
+type SubscriptionOptions struct {
+	// MaxBatchSize caps the maximum batch sized used when retrieving messages. It defaults to 1000.
```
sized -> size

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -387,19 +407,34 @@ func (*topic) ErrorCode(err error) gcerrors.ErrorCode {
 	path := fmt.Sprintf("projects/%s/subscriptions/%s", projectID, subscriptionName)
-	return pubsub.NewSubscription(openSubscription(client, path), nil, ackBatcherOpts)
+
+	if opts == nil {
```
To make sure we're always getting the default as documented, and to share code better between the two constructors, how about:

```
dsub := openSubscription(client, path, opts)
recvOpts := *defaultRecvBatcherOpts
recvOpts.MaxBatchSize = dsub.opts.MaxBatchSize
return pubsub.NewSubscription(dsub, recvOpts, ackBatcherOpts)
```

here and in `OpenSubscriptionByPath`.

Then in `openSubscription`:

```
if opts == nil {
  opts = &SubscriptionOptions{}
}
if opts.MaxBatchSize == 0 {
  opts.MaxBatchSize = defaultRecvBatchOpts.MaxBatchSize
}
return &subscription{client, subscriptionPath, opts}
```

I think for that to work you'll need to change the return value for `openSubscription` to `*subscription`, but that should be fine.
?

davidsbond replied:
Yep, this makes sense to me

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -412,12 +447,27 @@ func OpenSubscriptionByPath(client *raw.SubscriberClient, subscriptionPath strin
 	}
-	return pubsub.NewSubscription(openSubscription(client, subscriptionPath), nil, ackBatcherOpts), nil
+
+	if opts == nil {
```
See above.

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -525,3 +575,11 @@ func (*subscription) ErrorCode(err error) gcerrors.ErrorCode {
+		return 0, fmt.Errorf("expected only one parameter value, got: %v", len(value))
+	}
+
+	return strconv.ParseInt(value[0], 10, 64)
```
Maybe use `strconv.Atoi` here and return an `int` instead of `int64`?

vangent commented on pubsub/gcppubsub/gcppubsub_test.go:
```diff
@@ -408,6 +408,8 @@ func TestOpenSubscriptionFromURL(t *testing.T) {
 		// Invalid parameter.
 		{"gcppubsub://myproject/mysub?param=value", true},
+		// Valid parameters
+		{"gcppubsub://projects/myproject/subscriptions/mysub?max-batch-size=1", false},
```
Can you add a test with an invalid value for the new parameter?

davidsbond commented:
@vangent thanks for the second round of review. I think I've implemented everything as you've suggested. Let me know if there's more :)

vangent commented on pubsub/gcppubsub/gcppubsub.go:
```diff
@@ -212,13 +216,12 @@ func (o *URLOpener) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic
-		MaxBatchSize: defaultRecvBatcherOpts.MaxBatchSize,
-	}
+	opts := o.SubscriptionOptions
+	opts.MaxBatchSize = defaultRecvBatcherOpts.MaxBatchSize
```
Shouldn't need this (and in fact don't want it, in case there is a custom `URLOpener` with it set we don't want to overwrite it). If the value is 0, it will get the default via openSubscriber later.

davidsbond commented:
@vangent I rebased onto master and made your change

davidsbond commented:
@vangent thanks for the reviews! Looking forward to this being released, been a pleasure to contribute.

vangent commented:
No problem, and thanks for the contribution. You can use it right now if you like, by referencing the commit:
https://golang.org/ref/mod#vcs-pseudo
